src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseBaseKernels.hpp

```diff
@@ -464,6 +467,153 @@ struct SolutionCheckKernel
 
 };
 
+/******************************** HydrostaticPressureKernel ********************************/
+
+struct HydrostaticPressureKernel
+{
+  template< typename FLUID_WRAPPER >
+  static integer
+  launch( localIndex const size,
+          localIndex const numComps,
+          localIndex const numPhases,
+          localIndex const maxNumEquilIterations,
+          real64 const equilTolerance,
+          real64 const (&gravVector)[ 3 ],
+          real64 const & minElevation,
+          real64 const & dElevation,
+          real64 const & datumElevation,
+          real64 const & datumPres,
+          FLUID_WRAPPER fluidWrapper,
+          arrayView1d< TableFunction::KernelWrapper const > compFracTableWrappers,
+          TableFunction::KernelWrapper tempTableWrapper,
+          arrayView1d< arrayView1d< real64 > const > elevationValues,
+          arrayView1d< arrayView1d< real64 > const > pressureValues )
+  {
+    localIndex constexpr MAX_NUM_COMP = constitutive::MultiFluidBase::MAX_NUM_COMPONENTS;
+    localIndex constexpr MAX_NUM_PHASE = constitutive::MultiFluidBase::MAX_NUM_PHASES;
+
+    // Step 1: compute the mass density at the datum elevation
+    real64 const datumTemp = tempTableWrapper.compute( &datumElevation );
+    stackArray1d< real64, MAX_NUM_COMP > datumCompFrac( numComps );
+    stackArray1d< real64, MAX_NUM_PHASE > datumPhaseFrac( numPhases );
+    stackArray1d< real64, MAX_NUM_PHASE > datumPhaseDens( numPhases );
+    stackArray1d< real64, MAX_NUM_PHASE > datumPhaseMassDens( numPhases );
+    stackArray1d< real64, MAX_NUM_PHASE > datumPhaseVisc( numPhases );
+    stackArray2d< real64, MAX_NUM_COMP *MAX_NUM_PHASE > datumPhaseCompFrac( numPhases, numComps );
+    real64 datumTotalDens = 0.0;
+    for( localIndex ic = 0; ic < numComps; ++ic )
+    {
+      datumCompFrac[ic] = compFracTableWrappers[ic].compute( &datumElevation );
+    }
+
+    // TODO: find a way to make this compile on GPU despite the change in permutation
```

Comment:
We could do the following:
1. Add `PERMUTATION` template parameter to all `stackArrayNd` template aliases (make it trailing and defaulted).
2. Add an extra dimension (of size 1) and the expected layout to all these arrays:
```suggestion
    stackArray2d< real64, MAX_NUM_COMP, compflow::LAYOUT_COMP > datumCompFrac( 1, numComps );
    stackArray2d< real64, MAX_NUM_PHASE, multifluid::LAYOUT_PHASE > datumPhaseFrac( 1, numPhases );
    stackArray2d< real64, MAX_NUM_PHASE, multifluid::LAYOUT_PHASE > datumPhaseDens( 1, numPhases );
    stackArray2d< real64, MAX_NUM_PHASE, multifluid::LAYOUT_PHASE > datumPhaseMassDens( 1, numPhases );
    stackArray2d< real64, MAX_NUM_PHASE, multifluid::LAYOUT_PHASE > datumPhaseVisc( 1, numPhases );
    stackArray3d< real64, MAX_NUM_COMP *MAX_NUM_PHASE, multifluid::LAYOUT_PHASE_COMP > datumPhaseCompFrac( 1, numPhases, numComps );
    real64 datumTotalDens = 0.0;
    for( localIndex ic = 0; ic < numComps; ++ic )
    {
      datumCompFrac[0][ic] = compFracTableWrappers[ic].compute( &datumElevation );
    }

```

The alternative is ugly... making all fluid `KernelWrapper::compute` methods templated on types of all arguments, which might be unreasonably expensive in terms of compile time.

Reply:
You could also just make these regular arrays instead of stack (but keep the layout and extra dim). This might even be better, since you're capturing them in a GPU kernel below. Note though, that you need to capture a view, not the array itself - this can be done with an explicit capture list of the form
```
[=, 
 datumCompFrac = datumCompFrac.toViewConst(), 
 datumPhaseFrac = datumPhaseFrac.toViewConst(),
 ... ]
```

Reply:
Actually, the kernel below should probably remain on the host, since it's only computing a relatively small pressure table (it took me a while to notice that)

Reply:
Thanks very much @klevzoff for the suggestions. 

1. For the size of the table, it will always be small (less than 1000 entries). In the current version of the PR there is default number of entries in the table, but ultimately this should become a default spacing (dz). According to the documentation, the default spacing in IX is 2ft, so we can probably keep this kernel with `parallelHostPolicy`.
2. Regarding the array permutations, I implemented what you suggested. I added extra dimensions to the regular arrays with the layout, and I passed a view on the array needed in the kernel (`datumPhaseMassDensViewConst`) in case we change our mind and later decide to switch to `parallelDevicePolicy`.

Reply:
@klevzoff I ended up changing a few things in the initialization procedure, which now requires `serialPolicy` to work properly (it is now a serial algorithm initializing the pressure from the datum elevation). I also implemented the 2ft default spacing in the pressure table, which will therefore always be small. Now everything seems to work well. Thanks for your help!